### PR TITLE
feat(new codecs): introduce avro

### PR DESCRIPTION
**Pull request by Ion-manden**
This pr. is a continuation of #18636 by [@WarmSnowy](https://github.com/WarmSnowy), if they decide to return to it this can be closed in favor of theirs.
Thanks to [@cameronbraid](https://github.com/cameronbraid) and [@WarmSnowy](https://github.com/WarmSnowy) for the great work, I am just here to see if I can help get it across the line.

- Return error on Avro decimal, ducation and local timestamps
- Add comment to docs on unsupported types

### CODE REVIEW

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:245]: Nit 🥜: This is repeated across all tests, can you refactor it?

File: lib/codecs/tests/data/avro/README.md
Comment (pront) [inline lib/codecs/tests/data/avro/README.md:12]:
```suggestion
cargo run --package codecs --bin generate-avro-fixtures
```

**Review by pront — COMMENTED** [review]
Thank you for picking this up! We are almost there, left a few comments that should be easy to address.

**Comment by pront** [comment]
To reduce back-and-forth due to CI failures run the following before pushing commits:
```
# cd vector
cargo vdev fmt
cargo vdev check rust --clippy
cargo vdev test
```

More details can be found [here](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#running-other-checks).

File: lib/codecs/src/decoding/format/avro.rs
Comment (Ion-manden) [reply lib/codecs/src/decoding/format/avro.rs]: Ye of course, have moved it out so they can share, had to chage one to use message instead of uuid but tests are passing fine with it

File: lib/codecs/tests/bin/generate-avro-fixtures.rs
Comment (Ion-manden) [inline lib/codecs/tests/bin/generate-avro-fixtures.rs:178]: @pront  This is failing `cargo vdev check rust --clippy` with `error: name `SPADES` contains a capitalized acronym`

Should i just change the symbols to this error does not get generated? Or should i look for a way to ignore it?

Comment (pront) [reply lib/codecs/tests/bin/generate-avro-fixtures.rs:178]: You need to switch to camel case. see https://rust-lang.github.io/api-guidelines/naming.html.

Comment (Ion-manden) [reply lib/codecs/tests/bin/generate-avro-fixtures.rs:178]: Ye but they are generated from this test
```
    let schema = r#"
    {
        "type": "record",
        "name": "test",
        "fields": [
            {"name": "enum_field", "type": "enum", "symbols" : ["SPADES", "HEARTS", "DIAMONDS", "CLUBS"]}
        ]
    }
    "#;
```

`lib/codecs/tests/bin/generate-avro-fixtures.rs` +162

File: lib/codecs/tests/avro.rs
Comment (Ion-manden) [reply lib/codecs/tests/avro.rs]: Not sure if i am misunderstanding something, but this list is the exclude list.

Have removed duraiton as it is no longer supported and the tests for it have been removed.

Enum and long is being tested and passes.

Trying to see if i can get some of the others to work so they don't have to be excluded, but looks like encoding for them is not implemented.

```
thread 'roundtrip_avro_fixtures::case_1::path_04_tests_data_avro_generated_date_avro' panicked at 'called `Result::unwrap()` on an `Err` value: Date expected, got Long', lib/codecs/tests/avro.rs:43:52
```

File: website/cue/reference/components/sources/base/amqp.cue
Comment (urseberry) [inline website/cue/reference/components/sources/base/amqp.cue:65]:
```suggestion
																For Avro datum encoded in Kafka messages, the bytes are prefixed with the schema ID.  Set this to true to strip the schema ID prefix.
```

File: website/cue/reference/components/sources/base/aws_kinesis_firehose.cue
Comment (urseberry) [inline website/cue/reference/components/sources/base/aws_kinesis_firehose.cue:68]:
```suggestion
																For Avro datum encoded in Kafka messages, the bytes are prefixed with the schema ID.  Set this to true to strip the schema ID prefix.
```

File: website/cue/reference/components/sources/base/aws_s3.cue
Comment (urseberry) [inline website/cue/reference/components/sources/base/aws_s3.cue:163]:
```suggestion
																For Avro datum encoded in Kafka messages, the bytes are prefixed with the schema ID.  Set this to true to strip the schema ID prefix.
```

File: website/cue/reference/components/sources/base/aws_sqs.cue
Comment (urseberry) [inline website/cue/reference/components/sources/base/aws_sqs.cue:158]:
```suggestion
																For Avro datum encoded in Kafka messages, the bytes are prefixed with the schema ID.  Set this to true to strip the schema ID prefix.
```

**Review by urseberry — COMMENTED** [review]
Same comment for the rest of the cue files.

File: website/cue/reference/components/sources/base/amqp.cue
Comment (Ion-manden) [reply website/cue/reference/components/sources/base/amqp.cue]: All are generated from lib/codecs/src/decoding/format/avro.rs - have fixed the casing now and generated cue files

File: lib/codecs/tests/bin/generate-avro-fixtures.rs
Comment (pront) [reply lib/codecs/tests/bin/generate-avro-fixtures.rs:178]: You should be able to rename them and run the generation script again.

**Review by urseberry — APPROVED** [review]

File: lib/codecs/tests/avro.rs
Comment (pront) [inline lib/codecs/tests/avro.rs:22]: In the implementation we return an error for:    

* `Decimal`
* `Duration`
* `LocalTimestampMillis`
* `LocalTimestampMicros`

In this exclusion list I see `date`, `decimal_var` and `fixed`. Are these failing or fall under unsupported? 

For example, we do support `AvroValue::Fixed` but if the test fails, it means there is a bug.

**Review by pront — CHANGES_REQUESTED** [review]
We are almost there, left one comment.

**Review by jszwedko — APPROVED** [review]
Approving just the UX bits by reviewing the generated documentation.

File: website/cue/reference/components/sources/base/stdin.cue
Comment (jszwedko) [inline website/cue/reference/components/sources/base/stdin.cue:8]: Actually, I missed that this should be nested one more level. I.e. it should be at `decoding.codec.avro` rather than just `decoding.avro`.

**Review by jszwedko — CHANGES_REQUESTED** [review]
Missed one bit.

File: lib/codecs/tests/avro.rs
Comment (Ion-manden) [reply lib/codecs/tests/avro.rs:22]: So this seems to be an issue converting from Vrl to Avro

So date is not returned as a Date, but as a Long. Same issue for time_millis.

timestamp_micros and timestamp_millis are longs in Avro but maped to timestamps for Vrl, so when trying to change them back to Avro they fail with unexpected type.

Not quite sure how to fix it other than finding a way to know what type to convert to in encode in `lib/codecs/src/encoding/format/avro.rs` - maybe we can set the schema for the `apache_avro::to_value` somehow so it knows to change timestamp to long agin.

I can set them as unsupported for now if you want to, then i can open a new pr if i find a way to fix this.

File: website/cue/reference/components/sources/base/stdin.cue
Comment (Ion-manden) [reply website/cue/reference/components/sources/base/stdin.cue:8]: But it is one the same level as protobuf, json and syslog.
There is also a created entry in `decoding.codec.type.enum`

File: lib/codecs/tests/avro.rs
Comment (pront) [reply lib/codecs/tests/avro.rs]: Let's mark them as unsupported to unblock this PR. Please include the info you shared here in the docs.

File: website/cue/reference/components/sources/base/stdin.cue
Comment (jszwedko) [reply website/cue/reference/components/sources/base/stdin.cue:8]: Ah, you are right, the other codec options _are_ at that level.

**Review by jszwedko — APPROVED** [review]

File: lib/codecs/tests/avro.rs
Comment (Ion-manden) [reply lib/codecs/tests/avro.rs]: Returning error on them now, have added them to the unsupported list and shortly explained why with a link to my comment above.

File: lib/codecs/tests/bin/generate-avro-fixtures.rs
Comment (pront) [inline lib/codecs/tests/bin/generate-avro-fixtures.rs]: Let's not remove these since they might help future contributors. You can just tag them with `#[allow(unused)]`

File: lib/codecs/tests/avro.rs
Comment (pront) [inline lib/codecs/tests/avro.rs]: Let's add back the `exclude` since it makes it obvious which cases are unsupported.

File: lib/codecs/tests/bin/generate-avro-fixtures.rs
Comment (Ion-manden) [reply lib/codecs/tests/bin/generate-avro-fixtures.rs]: Sounds good, so keep the functions but not use them in the function that generates the files?

Comment (pront) [reply lib/codecs/tests/bin/generate-avro-fixtures.rs]: Yes, and use `#[allow(unused)]` to make clippy happy.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:110]:
```suggestion
```

I would move the rest of the details under the relevant match block in `try_from`:
```
    /// Example:
    /// timestamp_micros and timestamp_millis are longs in Avro but converted to timestamps for Vrl,
    /// so when trying to change them back to Avro they fail with unexpected type.
    /// This should be fixable by extending encode in `lib/codecs/src/encoding/format/avro.rs`
    /// to know the avro schema at serialization time.
    /// [See comment here](https://github.com/vectordotdev/vector/pull/19342#discussion_r1424215373)
```

Comment (Ion-manden) [reply lib/codecs/src/decoding/format/avro.rs:110]: Hmm, was just thinking that we could basically set them as i64 instead of parsing them as vrl timestamps.

Then leave it up to the user to convert in to timestamps using vrl.

Just changed it and the test parse.

```
        AvroValue::TimestampMicros(ts_micros) => Ok(VrlValue::from(ts_micros)),
        AvroValue::TimestampMillis(ts_millis) => Ok(VrlValue::from(ts_millis)),
```

Comment (Ion-manden) [reply lib/codecs/src/decoding/format/avro.rs:110]: Hmm we could also do the same for date, it just expects users to know how their avro works, but think that is a fair assumption.

Comment (Ion-manden) [reply lib/codecs/src/decoding/format/avro.rs]: [I added support for timestamps and local timestamps](https://github.com/vectordotdev/vector/pull/19342/commits/a669d7373a3bd404200c73922c813eac2cde3b54)

If you prefer to have it as timestamps in vrl i can revert this commit

Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs]: What does the encoder do? Is the roundtrip lossless?

Comment (Ion-manden) [reply lib/codecs/src/decoding/format/avro.rs]: The avro libs own avro to json conversion maps to number in json
```
            Value::TimestampMillis(t) => Ok(Self::Number(t.into())),
            Value::TimestampMicros(t) => Ok(Self::Number(t.into())),
            Value::LocalTimestampMillis(t) => Ok(Self::Number(t.into())),
            Value::LocalTimestampMicros(t) => Ok(Self::Number(t.into())),
```

Have added roundtrip test for local timestamp and timestamp kept the original timestamp one, both pass.

**Review by pront — APPROVED** [review]
Thank you @Ion-manden and @WarmSnowy for this big contribution! 

Do you have a list of future work items such as caveats with the current implementation or potential improvements.

**Pull request merged**
